libs/langchain/tests/unit_tests/agents/test_agent_async.py

```diff
@@ -0,0 +1,237 @@
+"""Unit tests for agents."""
+
+from typing import Any, Dict, List, Optional
+
+import pytest
+
+from langchain.agents import AgentExecutor, AgentType, initialize_agent
+from langchain.agents.tools import Tool
+from langchain.callbacks.manager import CallbackManagerForLLMRun
+from langchain.llms.base import LLM
+from tests.unit_tests.callbacks.fake_callback_handler import FakeCallbackHandler
+
+
+class FakeListLLM(LLM):
+    """Fake LLM for testing that outputs elements of a list."""
+
+    responses: List[str]
+    i: int = -1
+
+    def _call(
+        self,
+        prompt: str,
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> str:
+        """Increment counter, and then return response in that index."""
+        self.i += 1
+        print(f"=== Mock Response #{self.i} ===")
+        print(self.responses[self.i])
+        return self.responses[self.i]
+
+    def get_num_tokens(self, text: str) -> int:
+        """Return number of tokens in text."""
+        return len(text.split())
+
+    async def _acall(self, *args: Any, **kwargs: Any) -> str:
+        return self._call(*args, **kwargs)
+
+    @property
+    def _identifying_params(self) -> Dict[str, Any]:
+        return {}
+
+    @property
+    def _llm_type(self) -> str:
+        """Return type of llm."""
+        return "fake_list"
+
+
+def _get_agent(**kwargs: Any) -> AgentExecutor:
+    """Get agent for testing."""
+    bad_action_name = "BadAction"
+    responses = [
+        f"I'm turning evil\nAction: {bad_action_name}\nAction Input: misalignment",
+        "Oh well\nFinal Answer: curses foiled again",
+    ]
+    fake_llm = FakeListLLM(cache=False, responses=responses)
+
+    tools = [
+        Tool(
+            name="Search",
+            func=lambda x: x,
+            description="Useful for searching",
+        ),
+        Tool(
+            name="Lookup",
+            func=lambda x: x,
+            description="Useful for looking up things in a table",
+        ),
+    ]
+
+    agent = initialize_agent(
+        tools,
+        fake_llm,
+        agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION,
+        verbose=True,
+        **kwargs,
+    )
+    return agent
+
+
+@pytest.mark.asyncio
+async def test_agent_bad_action() -> None:
+    """Test react chain when bad action given."""
+    agent = _get_agent()
+    output = await agent.arun("when was langchain made")
```

Comment:
This test is a bit magical i think relies on knowledge of internal code too much? Could be helpful to document how the test works (why are 2 responses OK), once tool invocation fails what happens? another llm call that yields final answer?

Reply:
I just copied the existing sync test